### PR TITLE
Pivot table generation fails when no Columns and multiple Data are provided.

### PR DESCRIPTION
**Pull request by jinhyuk-kim-ca**
# PR Details
Hi,

I've been using excelize and found pivot table generation does not work when multiple Data are provided and no Columns are provided. Please see the detail below.

## Description
Here is a pivot table that I've been trying to generate with excelize.
image

And the base data for pivot table

Month | Year | Type | Sales | Sales1 | Sales2 | Sales3 | Sales4 | Sales5 | Sales6 | Region
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Jun | 2017 | Produce | 4059 | 2081 | 1318 | 4425 | 2540 | 456 | 3300 | North
Aug | 2017 | Dairy | 4728 | 3274 | 1211 | 1445 | 3237 | 4106 | 495 | North
Sep | 2018 | Produce | 4947 | 3287 | 2888 | 2790 | 3015 | 541 | 408 | South
Dec | 2017 | Meat | 1737 | 631 | 1485 | 26 | 1413 | 3090 | 194 | South
Feb | 2017 | Beverages | 4324 | 1159 | 1353 | 1957 | 3721 | 2189 | 2199 | East
Feb | 2019 | Beverages | 4703 | 4355 | 2451 | 3510 | 2605 | 156 | 3266 | East
Jun | 2019 | Produce | 746 | 1563 | 4376 | 4002 | 4718 | 447 | 94 | West
Aug | 2018 | Meat | 3623 | 953 | 1137 | 3133 | 4241 | 59 | 3033 | South
Dec | 2018 | Beverages | 4336 | 2546 | 4107 | 2940 | 1503 | 552 | 4843 | West
Mar | 2017 | Produce | 1515 | 4757 | 3687 | 3010 | 3410 | 285 | 3590 | East
Jul | 2018 | Produce | 3582 | 384 | 1297 | 4267 | 1137 | 4271 | 894 | North
Nov | 2017 | Produce | 2066 | 1270 | 493 | 3086 | 4819 | 3981 | 1052 | South
Feb | 2019 | Produce | 3749 | 1528 | 2818 | 4384 | 2903 | 1224 | 4547 | East
May | 2019 | Produce | 540 | 786 | 2051 | 3076 | 3640 | 2351 | 3844 | East
Jun | 2018 | Dairy | 90 | 1602 | 2258 | 3767 | 3231 | 2578 | 3154 | North
Apr | 2017 | Meat | 2743 | 1968 | 1166 | 3710 | 4535 | 440 | 4904 | North
Feb | 2017 | Produce | 3039 | 3430 | 4513 | 4700 | 1359 | 1720 | 783 | North
Jan | 2018 | Beverages | 565 | 4162 | 1829 | 2920 | 2048 | 1756 | 695 | North
Sep | 2019 | Dairy | 1092 | 3831 | 2577 | 2886 | 320 | 399 | 1162 | South
Jan | 2018 | Produce | 4103 | 4888 | 3318 | 3756 | 2019 | 2807 | 1157 | East
Apr | 2018 | Dairy | 3795 | 417 | 1393 | 3470 | 3996 | 4386 | 2632 | East
Sep | 2018 | Dairy | 1661 | 60 | 2420 | 1079 | 2954 | 1464 | 2060 | South
Oct | 2019 | Meat | 600 | 2039 | 4637 | 1533 | 561 | 2804 | 1685 | West
Aug | 2019 | Beverages | 1040 | 3662 | 740 | 1000 | 4284 | 3173 | 235 | South
Jun | 2017 | Beverages | 2869 | 3070 | 1336 | 3338 | 3472 | 1544 | 2395 | North
Oct | 2017 | Dairy | 1606 | 4648 | 3352 | 2420 | 3622 | 371 | 1117 | South
Jun | 2018 | Produce | 1592 | 129 | 4231 | 3721 | 855 | 4511 | 343 | West
Jul | 2017 | Produce | 3408 | 2008 | 528 | 3653 | 1774 | 3284 | 2375 | North
Nov | 2018 | Beverages | 4467 | 1114 | 4658 | 259 | 3546 | 4723 | 1724 | South
Feb | 2018 | Produce | 1531 | 3146 | 4983 | 928 | 516 | 4926 | 2477 | East

In case you need to sum of Sales, Sales1, Sales2... by "Type"(rows), The code would be like below,
```go
if err := f.AddPivotTable(&excelize.PivotTableOption {
                DataRange:       "Example!$A$1:$K$31",
                PivotTableRange: "Example!$A$32:$K$65",
		Rows: []excelize.PivotTableField{
			{Data: "Type"}},
		Filter:  []excelize.PivotTableField{},
		Columns: []excelize.PivotTableField{},
		Data: []excelize.PivotTableField{
			{Data: "Sales", Name: "Sum of Sales", Subtotal: "Sum"},
			{Data: "Sales1", Name: "Sum of Sales1", Subtotal: "Sum"},
			{Data: "Sales2", Name: "Sum of Sales2", Subtotal: "Sum"},
			{Data: "Sales3", Name: "Sum of Sales3", Subtotal: "Sum"},
			{Data: "Sales4", Name: "Sum of Sales4", Subtotal: "Sum"},
			{Data: "Sales5", Name: "Sum of Sales5", Subtotal: "Sum"},
			{Data: "Sales6", Name: "Sum of Sales6", Subtotal: "Sum"},
		},
		RowGrandTotals: true,
		ColGrandTotals: true,
		ShowDrill:      true,
		ShowRowHeaders: true,
		ShowColHeaders: true,
		ShowLastColumn: true,
	}); err != nil {
		fmt.Println(err)
	}
```

but this won't generate proper pivot table and end up with pivot table removed in the generated excel.

After doing comparison of "\xl\pivotTables\pivotTable1.xml",
the one that is working(generated manually) got **colFields** generated like,
`<colFields count="1">
	<field x="-2"/>
</colFields>`

but the one by excelize doesn't. (only colItems were generated)

I am not sure If I am following the standard, but I couldn't find what's "\<field x="-2"\>" the value of x means. 

here is the addition in order to generate colFields when there is no Columns input
```go
func (f *File) addPivotColFields(pt *xlsxPivotTableDefinition, opt *PivotTableOption) error {
	pt.ColFields = &xlsxColFields{}
	if len(opt.Columns) == 0 {
		//in order to create pivot table in case there is no input from Columns 
		pt.ColFields.Count = 1
		pt.ColFields.Field = append(pt.ColFields.Field, &xlsxField{
			X: -2,
		})
		return nil
	}
```
## How Has This Been Tested
go test TestAddPivotTable

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by xuri — CHANGES_REQUESTED** [review]
Thanks for your PR, please do not commit the spreadsheet files generated by `go test`.

**Comment by jinhyuk-kim-ca** [comment]
Thanks for reviewing, I have removed files generated by go test.

File: pivotTable.go [hunk 1]
```diff
@@ -459,12 +459,16 @@ func inPivotTableField(a []PivotTableField, x string) int {
 // addPivotColFields create pivot column fields by given pivot table
 // definition and option.
 func (f *File) addPivotColFields(pt *xlsxPivotTableDefinition, opt *PivotTableOption) error {
+	pt.ColFields = &xlsxColFields{}
 	if len(opt.Columns) == 0 {
+		//in order to create pivot table in case there is no input from Columns 
+		pt.ColFields.Count = 1
+		pt.ColFields.Field = append(pt.ColFields.Field, &xlsxField{
+			X: -2,
```
Comment (xuri) [inline pivotTable.go:467]: This modification will cause Excel application crash when there are multiple pivot tables in a worksheet. I think we need to resolve this issue in other way.

**Comment by lbraconnier2** [comment]
This commit, break the fact to have multiple columns and multi data. 
Now we can have :  no columns and many data, but we can't have both of them (many columns and many data).

it still ok for one column, one data.

**Comment by jinhyuk-kim-ca** [comment]
> This commit, break the fact to have multiple columns and multi data.
> Now we can have : no columns and many data, but we can't have both of them (many columns and many data).
> 
> it still ok for one column, one data.

Hi, May I ask a sample of pivot table that you are trying to generate by excel not with this package? with many columns and many data.
The reason that I am asking this is because I couldn't find a way to generate a pivot table with many columns and many data.
here is a pivot table that I tried with sample data in excelize doc.
image
image

**Comment by lbraconnier2** [comment]
it's seem to be the same kind of problem,  
with this code 

```
if err := ef.AddPivotTable(&excelize.PivotTableOption{
		DataRange:       "Sheet1!$A$1:$T$" + strconv.Itoa(i),
		PivotTableRange: "TCD!$G$2:$M$34",
		Rows: []excelize.PivotTableField{{Data: "Designation", DefaultSubtotal: false},
			{Data: "Sous_reference", DefaultSubtotal: false},
			{Data: "Type_prev", DefaultSubtotal: false},
			{Data: "Commentaire", DefaultSubtotal: false}},
		Filter:         []excelize.PivotTableField{{Data: "Code_pays"}},
		Columns:        []excelize.PivotTableField{{Data: "Annee", DefaultSubtotal: false}, {Data: "Trimestre", DefaultSubtotal: false}, {Data: "Mois", DefaultSubtotal: false}},
		Data:           []excelize.PivotTableField{{Data: "Qte_real", Name: "QTE REALISE", Subtotal: "Sum"}, {Data: "Mnt_real", Name: "MNT REALISE", Subtotal: "Sum"}},
		RowGrandTotals: true,
		//ColGrandTotals: true,
		ShowDrill:      true,
		ShowRowHeaders: true,
		ShowColHeaders: true,
		ShowLastColumn: true,
	}); err != nil {
		fmt.Println(err)
	}

```
After doing comparison of "\xl\pivotTables\pivotTable1.xml",
the one that is working(generated manually) got colFields generated like,
```
<colFields count="4">
		<field x="5"/>
		<field x="6"/>
		<field x="7"/>
		<field x="-2"/>		 
	</colFields>
```
The one by exelize is : 
```
<colFields count="3">
		<field x="5"/>
		<field x="6"/>
		<field x="7"/>	 
	</colFields>
```

**Comment by jinhyuk-kim-ca** [comment]
I looks like the issue you are having is related to the following lines.
https://github.com/360EntSecGroup-Skylar/excelize/blob/c49222023756d6740877f286cb7d3e7a5a0950eb/pivotTable.go#L478

The commit for this PR was before the line pivotTable.go#L478, when only Columns length is zero.

**Comment by lbraconnier2** [comment]
> Hi, May I ask a sample of pivot table that you are trying to generate by excel not with this package? with many columns and many data.
> The reason that I am asking this is because I couldn't find a way to generate a pivot table with many columns and many data.
> here is a pivot table that I tried with sample data in excelize doc.
> image
> image

[exemple.xlsx](https://github.com/360EntSecGroup-Skylar/excelize/files/5301238/exemple.xlsx)

**Pull request merged**